File: examples/example4/example4.cpp

#include "example4.hpp"

namespace example4 {

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

Start::~Start() noexcept {}
Name::~Name() noexcept {}
Str::~Str() noexcept {}


SimpleHello::SimpleHello(  )
  :std::tuple<  >(  ) {}

void SimpleHello::accept( Visitor& visitor ) {
  visitor.visitSimpleHello( *this );
}
void SimpleHello::accept( ConstVisitor& visitor ) const {
  visitor.visitSimpleHello( *this );
}


HelloWithName::HelloWithName( Name const& arg1 )
  :std::tuple< Name >( arg1 ) {}

void HelloWithName::accept( Visitor& visitor ) {
  visitor.visitHelloWithName( *this );
}
void HelloWithName::accept( ConstVisitor& visitor ) const {
  visitor.visitHelloWithName( *this );
}


NameString::NameString( Str const& arg1 )
  :std::tuple< Str >( arg1 ) {}

void NameString::accept( Visitor& visitor ) {
  visitor.visitNameString( *this );
}
void NameString::accept( ConstVisitor& visitor ) const {
  visitor.visitNameString( *this );
}



///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

std::ostream& operator <<( std::ostream &out, Start const& self ) {
  class Visitor : public Start::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitSimpleHello( SimpleHello const& host ) {
      *out_ << host;
    }
    void visitHelloWithName( HelloWithName const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, Name const& self ) {
  class Visitor : public Name::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitNameString( NameString const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, Str const& self ) {
  class Visitor : public Str::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
  } visitor( out );
  self.accept( visitor );
  return out;
}


std::ostream& operator <<( std::ostream& out, SimpleHello const& self ) {
  out << "SimpleHello(" << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, HelloWithName const& self ) {
  out << "HelloWithName("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, NameString const& self ) {
  out << "NameString("<<  std::get< 0 >( self ) << ")";
  return out;
}


///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

Node1::Node1() {}

Node2::Node2( Start const& content_ ) :content( content_ ) {}

Node3::Node3( Name const& content_ ) :content( content_ ) {}

Node4::Node4() {}

Node5::Node5( str const& content_ ) :content( content_ ) {}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

std::shared_ptr< State< Node1 > > begin() {
  std::shared_ptr< State<> > bottom( new State<>() );
  return State< Node1 >::make( Node1(), bottom );
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

}

File: examples/example4/Example4.hs

{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE UndecidableInstances #-}

module Example4 where

-------------------------------------------------------------------------------

-- grammar definition

-- simpleHello : Start -> "hello()"
-- helloWithName : Start -> "hello()" Name
-- nameString : Name -> str

-------------------------------------------------------------------------------

type Program r a = (a -> r) -> r

program :: a -> Program r a
program a = \k -> k a

-------------------------------------------------------------------------------

-- AST nodes

data Start
  = SimpleHello
  | HelloWithName Name
  deriving (Show)

data Name
  = NameString Str
  deriving (Show)

data Str
  deriving (Show)

-------------------------------------------------------------------------------

-- terminal symbols

class HelloTransition s t | s -> t where
  hello :: s -> Program r t

-------------------------------------------------------------------------------

-- automaton states

data Node1 prev = Node1 prev

data Node2 prev = Node2 prev Start

data Node3 prev = Node3 prev Name

data Node4 prev = Node4 prev

data Node5 prev = Node5 prev Str

-------------------------------------------------------------------------------

-- shift transitions

instance {-# OVERLAPS #-} HelloTransition (Node1 prev) (Node4 (Node1 prev)) where
  hello prev = program (Node4 prev)

-------------------------------------------------------------------------------

-- reduces

class Reduce s t | s -> t where
  reduce :: s -> t


-- simpleHello : Start -> "hello()"

instance Reduce (Node4 (Node1 prev)) (Node2 (Node1 prev)) where
  reduce (Node4 prev) = (Node2 prev (SimpleHello))


-- helloWithName : Start -> "hello()" Name

instance Reduce (Node3 (Node4 (Node1 prev))) (Node2 (Node1 prev)) where
  reduce (Node3 (Node4 prev) x1) = (Node2 prev (HelloWithName x1))


-- nameString : Name -> str

instance Reduce (Node5 (Node4 prev)) (Node3 (Node4 prev)) where
  reduce (Node5 prev x1) = (Node3 prev (NameString x1))


-- Reduce -> Transition

instance {-# OVERLAPS #-} (Reduce r s, HelloTransition s t) => HelloTransition r t where
  hello r = hello (reduce r)

-------------------------------------------------------------------------------

-- initial state

begin :: Program r (Node1 ())
begin = program (Node1 ())

-------------------------------------------------------------------------------

class End s r | s -> r where
  end :: s -> r

instance {-# OVERLAPS #-} End (Node2 prev) Start where
  end (Node2 _ r) = r

instance {-# OVERLAPS #-} (Reduce r s, End s t) => End r t where
  end r = end (reduce r)

-------------------------------------------------------------------------------

